enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -5821,31 +5940,54 @@ class AdjointGenerator
           return;
 
-        IRBuilder<> Builder2(call.getParent());
-        getReverseBuilder(Builder2);
-        Value *x = lookup(gutils->getNewFromOriginal(orig->getArgOperand(0)),
-                          Builder2);
-
-        SmallVector<Value *, 1> args = {x};
-        auto sinhf = gutils->oldFunc->getParent()->getOrInsertFunction(
-            (funcName == "sinh") ? "cosh" : "coshf", called->getFunctionType(),
-            called->getAttributes());
-        auto cal = cast<CallInst>(Builder2.CreateCall(sinhf, args));
-        Value *dif0 = Builder2.CreateFMul(diffe(orig, Builder2), cal);
-        setDiffe(orig, Constant::getNullValue(orig->getType()), Builder2);
-        addToDiffe(orig->getArgOperand(0), dif0, Builder2, x->getType());
-        return;
+        switch (Mode) {
+        case DerivativeMode::ForwardMode: {
+          IRBuilder<> Builder2(&call);
+          getForwardBuilder(Builder2);
+          Value *x = gutils->getNewFromOriginal(orig->getArgOperand(0));
+
+          SmallVector<Value *, 1> args = {x};
+          auto sinhf = gutils->oldFunc->getParent()->getOrInsertFunction(
+              (funcName == "sinh") ? "cosh" : "coshf",
+              called->getFunctionType(), called->getAttributes());
+          auto cal = cast<CallInst>(Builder2.CreateCall(sinhf, args));
+          Value *dif0 =
+              Builder2.CreateFMul(diffe(orig->getArgOperand(0), Builder2), cal);
+          setDiffe(orig, dif0, Builder2);
+          return;
+        }
+        case DerivativeMode::ReverseModeGradient:
+        case DerivativeMode::ReverseModeCombined: {
+          IRBuilder<> Builder2(call.getParent());
+          getReverseBuilder(Builder2);
+          Value *x = lookup(gutils->getNewFromOriginal(orig->getArgOperand(0)),
+                            Builder2);
+
+          SmallVector<Value *, 1> args = {x};
+          auto sinhf = gutils->oldFunc->getParent()->getOrInsertFunction(
+              (funcName == "sinh") ? "cosh" : "coshf",
+              called->getFunctionType(), called->getAttributes());
+          auto cal = cast<CallInst>(Builder2.CreateCall(sinhf, args));
+          Value *dif0 = Builder2.CreateFMul(diffe(orig, Builder2), cal);
+          setDiffe(orig, Constant::getNullValue(orig->getType()), Builder2);
+          addToDiffe(orig->getArgOperand(0), dif0, Builder2, x->getType());
+          return;
+        }
+        case DerivativeMode::ReverseModePrimal: {
+          return;
+        }
+        }
       }
 
       // Functions that only modify pointers and don't allocate memory,
       // needs to be run on shadow in primal
       if (funcName == "_ZSt29_Rb_tree_insert_and_rebalancebPSt18_Rb_tree_"
                       "node_baseS0_RS_") {
-        if (Mode == DerivativeMode::ReverseModeGradient) {
+        if (Mode == DerivativeMode::ReverseModeGradient ||
+            Mode == DerivativeMode::ForwardMode) {
           eraseIfUnused(*orig, /*erase*/ true, /*check*/ false);
```

Comment:
This still needs to be run in forward mode, whose implementation should be essentially the same as the augmented primal.